streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -924,28 +924,69 @@ private StreamThread createStreamThread(final long cacheSizePerThread, final int
      * @return name of the added stream thread or empty if a new stream thread could not be added
      */
     public Optional<String> addStreamThread() {
-        synchronized (changeThreadCount) {
-            if (isRunningOrRebalancing()) {
-                final int threadIdx = getNextThreadIndex();
-                final long cacheSizePerThread = getCacheSizePerThread(threads.size() + 1);
+        if (isRunningOrRebalancing()) {
+            final int threadIdx;
+            final long cacheSizePerThread;
+            synchronized (changeThreadCount) {
+                threadIdx = getNextThreadIndex();
+                cacheSizePerThread = getCacheSizePerThread(threads.size() + 1);
                 resizeThreadCache(cacheSizePerThread);
-                final StreamThread streamThread = createStreamThread(cacheSizePerThread, threadIdx);
-                synchronized (stateLock) {
-                    if (isRunningOrRebalancing()) {
-                        streamThread.start();
-                        return Optional.of(streamThread.getName());
-                    } else {
-                        streamThread.shutdown();
+            }
+            final StreamThread streamThread = createStreamThread(cacheSizePerThread, threadIdx);
+
+            synchronized (stateLock) {
+                if (isRunningOrRebalancing()) {
+                    streamThread.start();
+                    return Optional.of(streamThread.getName());
+                } else {
+                    streamThread.shutdown();
+                    threads.remove(streamThread);
+                    resizeThreadCache(getCacheSizePerThread(threads.size()));
+                    return Optional.empty();
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Removes one stream thread out of the running stream threads from this Kafka Streams client.
+     *
+     * The removed stream thread is gracefully shut down. This method does not specify which stream
+     * thread is shut down.
+     *
+     * Since the number of stream threads decreases, the sizes of the caches in the remaining stream
+     * threads are adapted so that the sum of the cache sizes over all stream threads equals the total
+     * cache size specified in configuration {@code cache.max.bytes.buffering}.
+     *
+     * @return name of the removed stream thread or empty if a stream thread could not be removed because
+     *         no stream threads are alive
+     */
+    public Optional<String> removeStreamThread() {
+        if (isRunningOrRebalancing()) {
+            for (final StreamThread streamThread : threads) {
+                if (streamThread.isAlive()) {
+                    streamThread.shutdown();
+                    while (streamThread.state() != StreamThread.State.DEAD && !streamThread.getName().equals(Thread.currentThread().getName())) {
```

Comment:
removeStreamThread is supposed to be a blocking call until the shutdown is complete

Reply:
I take it the `!streamThread.getName().equals(Thread.currentThread().getName())` is in preparation for adding the `REPLACE_THREAD` enum -- if so, can you just leave a `//TODO` here for now and add this in the followup PR so we have relevant changes reviewed together?

Reply:
Ah right you mean if `removeThread()` is called from the exception handler. Thanks for clarifying

Reply:
Ok in that case I think my concerns from [here](https://github.com/apache/kafka/pull/9695#discussion_r537976391) apply as well: is it safe to immediately redistribute the memory from the cache if the removed thread hasn't necessarily closed yet? Definitely risks an OOM, but maybe that risk is low (or at least acceptable). But we definitely can't remove the thread from `threads` until it's shut down since that puts its thread id back on the menu, and the ids have to be unique. How should we handle this?

Reply:
I think we have to put the thread id back on the menu. Otherwise when we replace the thread we will not be able to use the same id. In the tests for `REPLACE_THREAD` there has been no problems with having the thread that is shutting down have the same name as the thread that is starting. But for a call to `removeThread` it is probably best to not resize until shutdown is complete

i think the OOM risk is acceptable. The thread shutdown is much simpler than a client one and should not be using more space and the new thread should not have any tasks assigned until the old thread leaves the group.

Reply:
> Ah right you mean if removeThread() is called from the exception handler. Thanks for clarifying

It doesn't have to be from the handler. A thread that calls `removeThread()` could easily be removing itself.

In which case your concern about the OOM might apply as well. I don't see a good way around it though .

Reply:
>As for swallowing the InterruptedException, it still makes me uncomfortable but maybe there was/is a good reason that we do it already. Or, maybe you're just the first person to notice that we're doing that.

I hope there is a good reason, but we don't seem to have a problem with it yet so....